File: src/services/logging.ts
/**
 * PHASE 8: Observability - Frontend Logging Service
 * Centralized logging, error tracking, performance monitoring
 */

export enum LogLevel {
  DEBUG = 'DEBUG',
  INFO = 'INFO',
  WARN = 'WARN',
  ERROR = 'ERROR',
  CRITICAL = 'CRITICAL'
}

export interface LogEntry {
  timestamp: string
  level: LogLevel
  message: string
  context?: Record<string, any>
  stackTrace?: string
  url?: string
  userAgent?: string
  userId?: number
}

export interface PerformanceMetric {
  name: string
  duration: number
  timestamp: string
  metadata?: Record<string, any>
}

class LoggingService {
  private logs: LogEntry[] = []
  private metrics: PerformanceMetric[] = []
  private maxLogs = 1000
  private maxMetrics = 500
  private remoteEndpoint = '/api/logs'

  /**
   * Log message at specified level
   */
  log(level: LogLevel, message: string, context?: Record<string, any>): void {
    const entry: LogEntry = {
      timestamp: new Date().toISOString(),
      level,
      message,
      context,
      url: window.location.href,
      userAgent: navigator.userAgent
    }

    this.logs.push(entry)
    if (this.logs.length > this.maxLogs) {
      this.logs.shift()
    }

    // Console output in development
    if (import.meta.env.DEV) {
      const color = this.getLevelColor(level)
      console.log(`%c[${level}]`, `color: ${color}; font-weight: bold;`, message, context)
    }

    // Send critical errors to backend
    if (level === LogLevel.ERROR || level === LogLevel.CRITICAL) {
      this.sendToBackend(entry)
    }
  }

  debug(message: string, context?: Record<string, any>): void {
    this.log(LogLevel.DEBUG, message, context)
  }

  info(message: string, context?: Record<string, any>): void {
    this.log(LogLevel.INFO, message, context)
  }

  warn(message: string, context?: Record<string, any>): void {
    this.log(LogLevel.WARN, message, context)
  }

  error(message: string, context?: Record<string, any>, stackTrace?: string): void {
    const entry = { timestamp: new Date().toISOString(), level: LogLevel.ERROR, message, context, stackTrace, url: window.location.href, userAgent: navigator.userAgent }
    this.logs.push(entry)
    this.sendToBackend(entry)
  }

  critical(message: string, context?: Record<string, any>): void {
    const entry = { timestamp: new Date().toISOString(), level: LogLevel.CRITICAL, message, context, url: window.location.href, userAgent: navigator.userAgent }
    this.logs.push(entry)
    this.sendToBackend(entry)
  }

  /**
   * Track performance metric
   */
  trackMetric(name: string, duration: number, metadata?: Record<string, any>): void {
    const metric: PerformanceMetric = {
      name,
      duration,
      timestamp: new Date().toISOString(),
      metadata
    }

    this.metrics.push(metric)
    if (this.metrics.length > this.maxMetrics) {
      this.metrics.shift()
    }

    // Log slow operations (>1s)
    if (duration > 1000) {
      this.warn(`Slow operation: ${name} took ${duration}ms`, { duration, name })
    }
  }

  /**
   * Measure async operation
   */
  async measure<T>(name: string, fn: () => Promise<T>, metadata?: Record<string, any>): Promise<T> {
    const start = performance.now()
    try {
      const result = await fn()
      const duration = performance.now() - start
      this.trackMetric(name, duration, metadata)
      return result
    } catch (error) {
      const duration = performance.now() - start
      this.trackMetric(name, duration, { ...metadata, error: String(error) })
      throw error
    }
  }

  /**
   * Track API call
   */
  trackApiCall(method: string, url: string, duration: number, status: number): void {
    this.trackMetric(`API ${method} ${url}`, duration, { method, url, status })
  }

  /**
   * Track page navigation
   */
  trackPageNavigation(from: string, to: string): void {
    this.info(`Navigation: ${from} → ${to}`, { from, to })
  }

  /**
   * Track user action
   */
  trackUserAction(action: string, target: string, metadata?: Record<string, any>): void {
    this.info(`User action: ${action}`, { action, target, ...metadata })
  }

  /**
   * Track error
   */
  trackError(error: Error, context?: Record<string, any>): void {
    this.error(error.message, { ...context, name: error.name }, error.stack)
  }

  /**
   * Get all logs
   */
  getLogs(level?: LogLevel, limit?: number): LogEntry[] {
    let filtered = this.logs
    if (level) {
      filtered = filtered.filter(l => l.level === level)
    }
    if (limit) {
      filtered = filtered.slice(-limit)
    }
    return filtered
  }

  /**
   * Get performance metrics
   */
  getMetrics(name?: string, limit?: number): PerformanceMetric[] {
    let filtered = this.metrics
    if (name) {
      filtered = filtered.filter(m => m.name.includes(name))
    }
    if (limit) {
      filtered = filtered.slice(-limit)
    }
    return filtered
  }

  /**
   * Export logs for analysis
   */
  exportLogs(format: 'json' | 'csv' = 'json'): string {
    if (format === 'json') {
      return JSON.stringify(this.logs, null, 2)
    } else {
      const headers = ['timestamp', 'level', 'message', 'context']
      const rows = this.logs.map(l => [
        l.timestamp,
        l.level,
        l.message,
        JSON.stringify(l.context || {})
      ])
      return [headers, ...rows].map(r => r.map(c => `"${String(c).replace(/"/g, '""')}"`).join(',')).join('\n')
    }
  }

  /**
   * Clear logs
   */
  clearLogs(): void {
    this.logs = []
  }

  /**
   * Clear metrics
   */
  clearMetrics(): void {
    this.metrics = []
  }

  /**
   * Send logs to backend
   */
  private async sendToBackend(entry: LogEntry): Promise<void> {
    try {
      const response = await fetch(this.remoteEndpoint, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(entry)
      })
      if (!response.ok) {
        console.error(`Failed to send log to backend: ${response.status}`)
      }
    } catch (error) {
      console.error('Error sending log to backend:', error)
    }
  }

  /**
   * Get color for log level
   */
  private getLevelColor(level: LogLevel): string {
    const colors: Record<LogLevel, string> = {
      [LogLevel.DEBUG]: '#7F8C8D',
      [LogLevel.INFO]: '#3498DB',
      [LogLevel.WARN]: '#F39C12',
      [LogLevel.ERROR]: '#E74C3C',
      [LogLevel.CRITICAL]: '#C0392B'
    }
    return colors[level] || '#000'
  }
}

export const logger = new LoggingService()

/**
 * Vue plugin for logging
 */
export function useLogging() {
  return {
    logger,
    log: logger.log.bind(logger),
    debug: logger.debug.bind(logger),
    info: logger.info.bind(logger),
    warn: logger.warn.bind(logger),
    error: logger.error.bind(logger),
    critical: logger.critical.bind(logger),
    trackMetric: logger.trackMetric.bind(logger),
    measure: logger.measure.bind(logger),
    trackApiCall: logger.trackApiCall.bind(logger),
    trackUserAction: logger.trackUserAction.bind(logger),
    trackError: logger.trackError.bind(logger)
  }
}

File: backend/app/routers/logging.py
"""
PHASE 8: Observability - Logging Endpoint
Receive logs from frontend and store for analysis
"""

from fastapi import APIRouter, Request, HTTPException, status
from pydantic import BaseModel
from datetime import datetime
from typing import Optional, List
from enum import Enum
import logging

logger = logging.getLogger(__name__)

class LogLevel(str, Enum):
    DEBUG = "DEBUG"
    INFO = "INFO"
    WARN = "WARN"
    ERROR = "ERROR"
    CRITICAL = "CRITICAL"


class LogEntry(BaseModel):
    timestamp: str
    level: LogLevel
    message: str
    context: Optional[dict] = None
    stackTrace: Optional[str] = None
    url: Optional[str] = None
    userAgent: Optional[str] = None
    userId: Optional[int] = None


class PerformanceMetric(BaseModel):
    name: str
    duration: float
    timestamp: str
    metadata: Optional[dict] = None


class LoggingService:
    """Service to collect and analyze logs from frontend"""
    
    logs: List[LogEntry] = []
    metrics: List[PerformanceMetric] = []
    max_logs = 10000
    max_metrics = 5000
    
    @classmethod
    def add_log(cls, entry: LogEntry) -> None:
        """Add log entry"""
        cls.logs.append(entry)
        if len(cls.logs) > cls.max_logs:
            cls.logs.pop(0)
        
        # Log critical errors to backend logger
        if entry.level in [LogLevel.ERROR, LogLevel.CRITICAL]:
            logger.error(f"Frontend {entry.level}: {entry.message}", extra={
                "context": entry.context,
                "url": entry.url,
                "stackTrace": entry.stackTrace
            })
    
    @classmethod
    def add_metric(cls, metric: PerformanceMetric) -> None:
        """Add performance metric"""
        cls.metrics.append(metric)
        if len(cls.metrics) > cls.max_metrics:
            cls.metrics.pop(0)
        
        # Alert on slow operations (>2s)
        if metric.duration > 2000:
            logger.warning(f"Slow operation: {metric.name} took {metric.duration}ms")
    
    @classmethod
    def get_logs(cls, level: Optional[LogLevel] = None, limit: int = 100) -> List[LogEntry]:
        """Get logs, optionally filtered by level"""
        filtered = cls.logs
        if level:
            filtered = [l for l in filtered if l.level == level]
        return filtered[-limit:]
    
    @classmethod
    def get_metrics(cls, name: Optional[str] = None, limit: int = 100) -> List[PerformanceMetric]:
        """Get metrics, optionally filtered by name"""
        filtered = cls.metrics
        if name:
            filtered = [m for m in filtered if name in m.name]
        return filtered[-limit:]
    
    @classmethod
    def get_stats(cls) -> dict:
        """Get logging statistics"""
        error_count = len([l for l in cls.logs if l.level == LogLevel.ERROR])
        critical_count = len([l for l in cls.logs if l.level == LogLevel.CRITICAL])
        
        avg_duration = 0
        if cls.metrics:
            avg_duration = sum(m.duration for m in cls.metrics) / len(cls.metrics)
        
        slow_operations = len([m for m in cls.metrics if m.duration > 2000])
        
        return {
            "total_logs": len(cls.logs),
            "error_count": error_count,
            "critical_count": critical_count,
            "total_metrics": len(cls.metrics),
            "avg_duration_ms": avg_duration,
            "slow_operations": slow_operations
        }
    
    @classmethod
    def clear_logs(cls) -> None:
        """Clear all logs"""
        cls.logs.clear()
    
    @classmethod
    def clear_metrics(cls) -> None:
        """Clear all metrics"""
        cls.metrics.clear()


router = APIRouter(prefix="/api", tags=["logging"])


@router.post("/logs")
async def log_event(entry: LogEntry, request: Request):
    """Receive log event from frontend"""
    try:
        # Add client IP
        entry.context = entry.context or {}
        entry.context["client_ip"] = request.client.host if request.client else "unknown"
        
        LoggingService.add_log(entry)
        logger.debug(f"Log received: {entry.level} - {entry.message}")
        
        return {"status": "ok"}
    except Exception as e:
        logger.error(f"Error processing log: {str(e)}")
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR)


@router.post("/metrics")
async def log_metric(metric: PerformanceMetric):
    """Receive performance metric from frontend"""
    try:
        LoggingService.add_metric(metric)
        logger.debug(f"Metric received: {metric.name} - {metric.duration}ms")
        
        return {"status": "ok"}
    except Exception as e:
        logger.error(f"Error processing metric: {str(e)}")
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR)


@router.get("/logs")
async def get_logs(level: Optional[LogLevel] = None, limit: int = 100):
    """Get logs from backend storage"""
    logs = LoggingService.get_logs(level, limit)
    return {"count": len(logs), "logs": logs}


@router.get("/metrics")
async def get_metrics(name: Optional[str] = None, limit: int = 100):
    """Get performance metrics"""
    metrics = LoggingService.get_metrics(name, limit)
    return {"count": len(metrics), "metrics": metrics}


@router.get("/logs/stats")
async def get_log_stats():
    """Get logging statistics"""
    return LoggingService.get_stats()


@router.delete("/logs")
async def clear_logs():
    """Clear all logs"""
    LoggingService.clear_logs()
    return {"status": "cleared"}


@router.delete("/metrics")
async def clear_metrics():
    """Clear all metrics"""
    LoggingService.clear_metrics()
    return {"status": "cleared"}
